### PR TITLE
[WOR-1309] Don't error for non-Ready MC Workspaces that have no cloud context.

### PR DESCRIPTION
**Pull request by cahrens**
Ticket: https://broadworkbench.atlassian.net/browse/WOR-1309

The particular use case is that we need to be able to query workspace details for a workspace that is deleting-- there is a period of time when WSM has already deleted the cloud context, but has not yet finished deleting the workspace itself.

We decided to generalize the change though to any workspace that is not in the Ready state.

---

**PR checklist**

- [ ] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala).
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala:33]: This is to handle the single usage by SnapshotService, at line https://github.com/broadinstitute/rawls/pull/2591/files#diff-3fb111d317a44913c11c6b540e6a2d05fdf5715c28ae0a0bb87921bb8e1057eeL80

Before the change `getCloudPlatform` would have thrown an exception for a workspace in this state.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:340]: Before this keyed off of `azureCloudContext`, returning bucket details if None. I have swapped the check so that we return bucket details if `googleProjectId` is not None, to handle the case of them both being None (plus it just makes more sense!).

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:357]: If this ends up being None, `getUserComputePermissions` will default to GCP permission model (Azure is slightly more lenient):

```
  private def getUserComputePermissions(workspaceId: String,
                                        userAccessLevel: WorkspaceAccessLevel,
                                        cloudPlatform: Option[WorkspaceCloudPlatform]
  ): Future[Boolean] =
    cloudPlatform match {
      case Some(WorkspaceCloudPlatform.Azure) =>
        Future.successful(userAccessLevel >= WorkspaceAccessLevels.Write)
      case _ if userAccessLevel >= WorkspaceAccessLevels.Owner =>
        Future.successful(true)
      case default =>
        samDAO.userHasAction(SamResourceTypeNames.workspace, workspaceId, SamWorkspaceActions.compute, ctx)
    }

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:431]: This could already handle None.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1010]: This could already handle None, but we will also filter out these workspaces below.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/AggregatedWorkspaceSpec.scala
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/AggregatedWorkspaceSpec.scala:89]: Checking that multiple cloud contexts throw an exception even for a non-ready workspace.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:3314]: Add in another Azure workspace without cloud context so we can catch both ways that the filtering happens (Ready workspace vs. non-Ready workspace).

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1018]: This is to maintain our previous behavior for the list endpoint… though I am starting to wonder if we could just allow workspaces with no cloud context to be returned. However, we would need to check that the list UI view can handle the cloud context being None, and think through if there are any other ramifications.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala:33]: FYI @jladieu ^

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1018]: That'd be good, if only to clean up the special casing on this codepath.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala
Comment (aherbst-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:3029]: I don't understand this comment, are you saying the workspace should not be returned?

File: core/src/main/resources/swagger/api-docs.yaml
Comment (aherbst-broad) [inline core/src/main/resources/swagger/api-docs.yaml:4137]: Good update

**Review by aherbst-broad — APPROVED** [review]

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala
Comment (cahrens) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:3029]: It was a comment that was already there… it's trying to point out that an azureContext is explicitly NOT being returned. If it were returned, it would look like this:

```
      when(service.workspaceManagerDAO.getWorkspace(azureWorkspace.workspaceIdAsUUID, services.ctx1))
        .thenReturn(
          new WorkspaceDescription()
            .stage(WorkspaceStageModel.MC_WORKSPACE)
            .azureContext(
              new AzureContext()
                .tenantId(UUID.randomUUID.toString)
                .subscriptionId(UUID.randomUUID.toString)
                .resourceGroupId(UUID.randomUUID.toString)
            )
        )

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala:33]: I'm having some trouble writing a test case for line 39. The test case that covers line 47 requires that a workspace actually exist (and it's in Ready state)-- it is using `testData.azureWorkspace`.

Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala:33]: Ah never mind, I figured it out. I just pushed up a test case.

**Comment by cahrens** [comment]
jenkins retest

**Review by blakery — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:1018]: Created https://broadworkbench.atlassian.net/browse/WOR-1314

**Comment by cahrens** [comment]
jenkins retest

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala
Comment (jladieu) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:86]: 1) If I understand your comments correctly, it sounds like `wsmWorkspace` will _always_ have `getCloudPlatform` if the workspace state is `READY`; otherwise `fetchAggregatedWorkspace` will throw an error.  Do I have that right?

2) I also think maybe there's a business logic rule that isn't codified here explicitly (correct me if I'm wrong though as I'm not sure if my hunch is correct), which is that we should only create snapshots for READY workspaces.

If both of my assertions are correct, then I wonder if you can avoid having to make any changes within `validateWorkspacePlatformCompatibility` by just having this try block check the WSM state is READY and fail for that reason (rather than later on due to the omission of the platform).

As long as both (1) and (2) are actually true, I think this might be a bit more intention revealing.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/AggregatedWorkspaceService.scala
Comment (jladieu) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/AggregatedWorkspaceService.scala:47]: Oof, this got pretty intense (scala newb here).  I defer to you on this as the code owner, but definitely whistled past this block on my way to the stuff I'm more familiar with.

**Review by jladieu — APPROVED** [review]
Approved with a comment/suggestion...I'm not actually sure if my suggestion is merited!  Please let me know either way so I can learn the logic.

Full disclosure, I honed in mainly on the stuff that's relevant to the place I was fiddling with last week, and defer judgment on the remainder to WSM folks!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala
Comment (cahrens) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:86]: Yes, getCloudPlatform will throw an error if a workspace is in the Ready state and has no CloudPlatform.

I agree with your second point, but I didn't know what the UI might do (or not) to prevent the user from trying to use a non-Ready workspace. Would you object to me submitting this as-is and possibly either you or I do a follow-up PR specific to the snapshot service? I would love to get this change in ASAP because it alleviates workspace deletion pain.

Comment (jladieu) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotService.scala:86]: Sounds good, go for it!

**Pull request merged**
